File: patch_workspace2.py
import re
with open('frontend/src/components/workspace/DocumentWorkspace.tsx', 'r') as f:
    content = f.read()

# Fix workspaceRef typeof function
content = content.replace("if (typeof workspaceRef === 'function') workspaceRef(node);", "if (typeof workspaceRef === 'function') (workspaceRef as any)(node);")

# Use renderSize or fallback to size
content = content.replace("style={{ width: size.width, minHeight: size.height }}", "style={{ width: renderSize?.width || size.width, minHeight: renderSize?.height || size.height }}")
content = content.replace("const pageWidth = size.width / scale;", "const pageWidth = (renderSize?.width || size.width) / scale;")
content = content.replace("const pageHeight = size.height / scale;", "const pageHeight = (renderSize?.height || size.height) / scale;")

# Fix canvasRef typing
content = content.replace("const canvasRef = React.useRef<HTMLCanvasElement | null>(null);", "const canvasRef = React.useRef<HTMLCanvasElement>(null);")

with open('frontend/src/components/workspace/DocumentWorkspace.tsx', 'w') as f:
    f.write(content)

File: patch_workspace2.js
const fs = require('fs');

const path = 'frontend/src/components/workspace/DocumentWorkspace.tsx';
let content = fs.readFileSync(path, 'utf8');

// Fix DocumentWorkspace.tsx(498,51): error TS2345: Argument of type 'number' is not assignable to parameter of type 'HTMLCanvasElement'.
// It seems `usePageRenderer({ doc, pageNumber, scale, canvasRef });` wasn't replacing the whole old effect block correctly or left an old render call somewhere.
content = content.replace("PdfRendererAdapter.renderPage(page, canvas, viewport);", "");
content = content.replace("PdfRendererAdapter.renderPage(page, canvas, viewport, {", "");
content = content.replace("  annotationMode: 0, // Disable built-in annotations", "");
content = content.replace("});", "");

fs.writeFileSync(path, content);
